blogposts/2023-02-10-adding-snapshot-testing-to-rnw-apps.md

```diff
@@ -0,0 +1,93 @@
+	3. React Native Gallery has a wide set of community modules it has tests for. Check out our [`jest-setup.js`](https://github.com/microsoft/react-native-gallery/blob/main/jest-setup.js) to see how we mocked needed modules. 
+
+Here is React Native Gallery's Jest customization within [`package.json`](https://github.com/microsoft/react-native-gallery/blob/main/package.json):
+```json
+"jest": {
+    "preset": "react-native",
+    "modulePathIgnorePatterns": [
+      "<rootDir>/lib/"
+    ],
+    "transformIgnorePatterns": [
+      "node_modules/(?!(react-native|react-native-xaml|@react-native|react-native-windows|react-native-config|@react-native-community|react-native-print|react-native-webview|react-native-windows-hello|react-native-permissions|react-native-tts|react-native-gesture-handler)/)"
+    ],
+    "setupFiles": [
+      "<rootDir>/jest-setup.js"
+    ]
+  },
+```
+
+Once you have all your tests running and passing, you might see that the Jest run is still exiting with exit code 1. This occurred in React Native Gallery. After some searching, I discovered it was because some of the components our tests were rendering contained asynchronous calls and updates. This results in Jest throwing errors because it cannot ensure that all tasks such as renders, user events, or data fetching were completed and applied before the snapshot match assertion was made. To resolve this error, you'll need to wrap the render of the component you wish to snapshot in React's `act()` function. Then, following the `act()` call, make your snapshot assertion. 
+
+Here's a code snippet of the test for React Native Gallery's `DateTimePicker` page:
+```js
+test('TimePicker Example Page', async () => {
+  let tree;
+  await act(() => {
+    tree = create(<TimePickerExamplePage />).toJSON();
+  });
+  expect(tree).toMatchSnapshot();
+});
+```
+
+A couple of recommendations:
+	1. The code for writing the tests is fairly straightforward; getting Jest configured correctly for your app is where the challenge kicks in. If you're working with an app with a number of community modules, tackle adding tests which use each community module one at a time.
+	2. Develop incrementally. Start with a test case which simply renders a `<View/>` control. Make sure your base case runs successfully. Then, test as you go. Jest errors aren't always clear, so the more granular you can make your changes between runs of the tests, the easier you'll be able to diagnose what's wrong.
+
+Additional Troubleshooting:
+	1. Attempting to run tests on the `react-native-permissions` module causes Jest to hang. As of now, we haven't solved this issue, so the test is disabled.
+	2. Some modules usage of the `UIManager.getViewManagerConfig()` API does not work well with Jest. `UIManager` returns undefined, even though the component is a part of core and thus should have already been transformed by Jest. As of now, we've had to disable tests for `TrackPlayer` and `SketchCanvas`.
```

Comment:
If this section is meant to be the conclusion, it misses a bit of the "skeleton" of a conclusion; I would suggest adding one or more phrases at the start like "Concluding, in this article I've shared {etc etc etc} and this has helped us reach {goals} - before leaving, I also wanted to are a couple of recommendations: [...] And here are a couple more pointers for some blockers I've hit while working on this".

Reply:
Will do!